mm2src/mm2_main/src/lp_swap/taker_swap.rs

```diff
@@ -1905,86 +1906,91 @@ impl TakerSwap {
+                        // We better find a way to rollback the state machine and remove erroneous events,
+                        // the swap at this point will be marked as errored but in fact it recovered from the error.
+                        return Ok((Some(TakerSwapCommand::ConfirmMakerPaymentSpend), vec![
+                            TakerSwapEvent::MakerPaymentSpent(tx_ident),
+                        ]));
+                    },
                 },
-            }
-        }
 
-        let other_taker_coin_htlc_pub = self.r().other_taker_coin_htlc_pub;
-        let secret_hash = self.r().secret_hash.clone();
-        let swap_contract_address = self.r().data.taker_coin_swap_contract_address.clone();
-        let watcher_reward = self.r().watcher_reward;
-        let refund_result = self
-            .taker_coin
-            .send_taker_refunds_payment(RefundPaymentArgs {
-                payment_tx: &taker_payment,
-                time_lock: locktime,
-                other_pubkey: other_taker_coin_htlc_pub.as_slice(),
-                tx_type_with_secret_hash: SwapTxTypeWithSecretHash::TakerOrMakerPayment {
-                    maker_secret_hash: &secret_hash,
+                // Encountered an error during swap recover.
+                Err(err) => match err {
+                    // The payment tx we want to refund isn't even on-chain. There is nothing to refund/spend.
+                    RecoverSwapError::PaymentTxNotFound => {
+                        return Ok((Some(TakerSwapCommand::Finish), vec![
+                            TakerSwapEvent::TakerPaymentRefundFailed(
+                                "TakerPayment isn't even on-chain to refund it.".into(),
+                            ),
+                        ]));
+                    },
+                    // The error is unrecoverable, retrying will not fix the issue.
+                    RecoverSwapError::Irrecoverable(e) => {
+                        return Ok((Some(TakerSwapCommand::Finish), vec![
+                            TakerSwapEvent::TakerPaymentRefundFailed(ERRL!("!taker_coin.recover_funds: {}", e).into()),
+                        ]));
+                    },
+                    // The error is temporary, retrying may fix the issue.
+                    RecoverSwapError::Temporary(e) => {
+                        error!("Error {} on recover_funds, retrying in 30 seconds", e);
+                        Timer::sleep(30.).await;
+                    },
+                    // We should wait for this many seconds and try again.
+                    RecoverSwapError::WaitAndRetry(secs) => {
+                        Timer::sleep(secs as f64).await;
+                    },
+                    // The swap will be automatically recovered after the specified locktime.
+                    RecoverSwapError::AutoRecoverableAfter(locktime) => {
+                        let taker_payment = self.r().taker_payment.clone().unwrap().tx_hex;
```

Comment:
Its better to use as_ref for taker_payment
```rust
let taker_payment = self.r().taker_payment.as_ref().unwrap().tx_hex.clone();
```

Reply:
[5e0709d](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2280/commits/5e0709d22a683c3304019872885aeaf438feb20f)